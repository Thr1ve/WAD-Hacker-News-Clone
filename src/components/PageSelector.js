import React, { PropTypes } from 'react';
import { rangeAround } from '../lib';

// NOTE: I want this to end up being an optional prop to the component instead
// of a constant; I'd like to find a way to have the component try
// to calculate this on its own by default.
const MAX_PAGE_SPACE = 7;

const PageSelector = ({ createClickHandler, prev, next, nPages = 1, currentPage = 1 }) =>
  <nav className="nav">
    <div className="pagination container">
      <a onClick={prev} className="button">Previous</a>
      <a onClick={next} className="button">Next</a>
      <ul>
        {
          rangeAround(currentPage, MAX_PAGE_SPACE).map(pageNumber =>
            <li key={pageNumber}>
              <a
                className={`button is-hidden-mobile${pageNumber === currentPage ? ' is-primary' : ''}`}
                onClick={createClickHandler(pageNumber)}
              > {pageNumber} </a>
            </li>
          )
        }
      </ul>
    </div>
  </nav>;

PageSelector.propTypes = {
  createClickHandler: PropTypes.func,
  prev: PropTypes.func,
  next: PropTypes.func,
  nPages: PropTypes.number,
  currentPage: PropTypes.number
}

export default PageSelector;
