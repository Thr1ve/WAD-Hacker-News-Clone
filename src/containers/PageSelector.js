import React, { PropTypes } from 'react';
import { connect } from 'react-redux';

import { nextPage, previousPage, setPage, } from '../actions';
import { getLastPage } from '../reducers';

import PageSelector from '../components/PageSelector';

const PageSelectorContainer = React.createClass({
  render() {
    return (
      <PageSelector
        createClickHandler={this.props.createClickHandler}
        prev={this.props.prev}
        next={this.props.next}
        nPages={this.props.nPages}
        currentPage={this.props.currentPage}
      />
    );
  }
})

function mapDispatchToProps(dispatch) {
  return {
    createClickHandler(pageNumber) {
      return function() {
        dispatch(setPage(pageNumber))
      }
    },
    next() { dispatch(nextPage()) },
    prev() { dispatch(previousPage()) },
  }
}

function mapStateToProps(state) {
  return {
    nPages: getLastPage(state),
    currentPage: state.ui.currentPage
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(PageSelectorContainer);
