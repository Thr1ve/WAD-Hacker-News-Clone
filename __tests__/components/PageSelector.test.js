import React from 'react';
import renderer from 'react-test-renderer';
import PageSelector from '../../src/components/PageSelector';

const onClick = () => {};
const createOnClick = () => onClick;
const clickHandlers = { createClickHandler: createOnClick, next: onClick, prev: onClick };

describe('PageSelector:snapshot', () => {
  it('should render correctly on page 1', () => {
    const component = renderer.create(
      <PageSelector nPages={10} currentPage={1} {...clickHandlers}/>
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly on page 2', () => {
    const component = renderer.create(
      <PageSelector nPages={10} currentPage={2} {...clickHandlers}/>
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly on page 3', () => {
    const component = renderer.create(
      <PageSelector nPages={10} currentPage={3} {...clickHandlers}/>
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly on page 4', () => {
    const component = renderer.create(
      <PageSelector nPages={10} currentPage={4} {...clickHandlers}/>
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly on page 5', () => {
    const component = renderer.create(
      <PageSelector nPages={10} currentPage={5} {...clickHandlers}/>
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
