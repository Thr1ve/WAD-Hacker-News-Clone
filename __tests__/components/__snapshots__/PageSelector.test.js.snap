exports[`PageSelector:snapshot should render correctly 1`] = `
<nav
  className="nav">
  <div
    className="pagination container">
    <a
      className="button"
      onClick={[Function noOp]}>
      Previous
    </a>
    <a
      className="button"
      onClick={[Function noOp]}>
      Next
    </a>
    <ul>
      <li>
        <a
          className="button is-hidden-mobile"
          onClick={[Function noOp]}>
           
          2
           
        </a>
      </li>
      <li>
        <a
          className="button is-hidden-mobile"
          onClick={[Function noOp]}>
           
          3
           
        </a>
      </li>
      <li>
        <a
          className="button is-hidden-mobile"
          onClick={[Function noOp]}>
           
          4
           
        </a>
      </li>
      <li>
        <a
          className="button is-hidden-mobile is-primary"
          onClick={[Function noOp]}>
           
          5
           
        </a>
      </li>
      <li>
        <a
          className="button is-hidden-mobile"
          onClick={[Function noOp]}>
           
          6
           
        </a>
      </li>
      <li>
        <a
          className="button is-hidden-mobile"
          onClick={[Function noOp]}>
           
          7
           
        </a>
      </li>
      <li>
        <a
          className="button is-hidden-mobile"
          onClick={[Function noOp]}>
           
          8
           
        </a>
      </li>
    </ul>
  </div>
</nav>
`;
